File: BeepingCoreTests/BeepingDelegateTests.m
@import XCTest;

#import "Beeping.h"
#import "BeepingEvent.h"

@interface Beeping (Testing)
- (void)onBeep:(NSNumber *)value;
- (BeepingEvent *)buildEventForToken:(int)token;
@end

@interface BeepingDelegateMock : NSObject <beepingDelegate>
@property (nonatomic) NSInteger beepIdCalls;
@property (nonatomic) NSInteger eventCalls;
@property (nonatomic, strong) BeepingEvent *lastEvent;
@property (nonatomic, strong) NSString *lastKey;
@end

@implementation BeepingDelegateMock

- (void)beepIdWith:(NSString *)beep_id {
    XCTAssertTrue([NSThread isMainThread]);
    self.beepIdCalls += 1;
    self.lastKey = beep_id;
}

- (void)beepingDidReceiveEvent:(BeepingEvent *)event {
    XCTAssertTrue([NSThread isMainThread]);
    self.eventCalls += 1;
    self.lastEvent = event;
}

@end

@interface StubBeepingCore : NSObject
@property (nonatomic, copy) NSString *decodedString;
@property (nonatomic, copy) NSString *decodedKey;
@property (nonatomic) NSInteger decodedMode;
@property (nonatomic) NSInteger decodedTimeStamp;
@property (nonatomic) float confidence;
@property (nonatomic) float confidenceError;
@property (nonatomic) float confidenceNoise;
@property (nonatomic) float receivedBeepsVolume;
@end

@implementation StubBeepingCore
- (NSString *)getDecodedString { return self.decodedString; }
- (NSString *)getDecodedKey { return self.decodedKey; }
- (int)getDecodedMode { return (int)self.decodedMode; }
- (int)getDecodedTimeStamp { return (int)self.decodedTimeStamp; }
- (float)getConfidence { return self.confidence; }
- (float)getConfidenceError { return self.confidenceError; }
- (float)getConfidenceNoise { return self.confidenceNoise; }
- (float)getReceivedBeepsVolume { return self.receivedBeepsVolume; }
@end

@interface BeepingDelegateTests : XCTestCase
@end

@implementation BeepingDelegateTests

- (Beeping *)freshBeepingWithStub:(StubBeepingCore *)stub delegate:(BeepingDelegateMock *)delegate {
    Beeping *beeping = [Beeping instance];
    [beeping setDelegate:delegate];
    [beeping setValue:stub forKey:@"_beepingCore"];
    return beeping;
}

- (void)testEndOkEmitsEventAndBeepId {
    StubBeepingCore *stub = [StubBeepingCore new];
    stub.decodedString = @"12345abcd";
    stub.decodedKey = @"12345";
    stub.decodedMode = 2;
    stub.decodedTimeStamp = 99;
    stub.confidence = 0.8f;
    stub.confidenceError = 0.1f;
    stub.confidenceNoise = 0.2f;
    stub.receivedBeepsVolume = -5.f;

    BeepingDelegateMock *delegate = [BeepingDelegateMock new];
    Beeping *beeping = [self freshBeepingWithStub:stub delegate:delegate];

    [beeping onBeep:@(BEEP_TOKEN_END_OK)];

    XCTAssertEqual(delegate.eventCalls, 1);
    XCTAssertEqual(delegate.beepIdCalls, 1);
    XCTAssertEqual(delegate.lastEvent.status, BeepingEventStatusEndOk);
    XCTAssertEqualObjects(delegate.lastEvent.key, @"12345");
    XCTAssertNotNil(delegate.lastEvent.decodedString);
}

- (void)testEndBadEmitsOnlyEvent {
    StubBeepingCore *stub = [StubBeepingCore new];
    stub.decodedString = @"99999abcd";
    stub.decodedMode = 1;
    stub.confidence = 0.2f;

    BeepingDelegateMock *delegate = [BeepingDelegateMock new];
    Beeping *beeping = [self freshBeepingWithStub:stub delegate:delegate];

    [beeping onBeep:@(BEEP_TOKEN_END_BAD)];

    XCTAssertEqual(delegate.eventCalls, 1);
    XCTAssertEqual(delegate.beepIdCalls, 0);
    XCTAssertEqual(delegate.lastEvent.status, BeepingEventStatusEndBad);
}

- (void)testStartEmitsEventOnly {
    StubBeepingCore *stub = [StubBeepingCore new];
    BeepingDelegateMock *delegate = [BeepingDelegateMock new];
    Beeping *beeping = [self freshBeepingWithStub:stub delegate:delegate];

    [beeping onBeep:@(BEEP_TOKEN_START)];

    XCTAssertEqual(delegate.eventCalls, 1);
    XCTAssertEqual(delegate.beepIdCalls, 0);
    XCTAssertEqual(delegate.lastEvent.status, BeepingEventStatusStart);
}

@end

File: BeepingCoreTests/BeepingDelegateTests.swift
//
//  BeepingDelegateTests.swift
//  BeepingTests
//
//  Swift Testing suite for `Beeping`'s delegate dispatch behavior.
//  Replaces the legacy `BeepingDelegateTests.m` (deleted in BEE-68 phase 5)
//  which used KVC reflection (`setValue:forKey:@"_beepingCore"`) to
//  inject a stub `BeepingCore`. Swift class internals are not
//  KVC-accessible; we use the internal test seam
//  `_injectEventForTesting(_:)` instead, which exercises the same dispatch
//  rules without needing to fake the audio pipeline.
//

import Testing
@testable import Beeping

/// Spy delegate for verifying which selectors fire and with what data.
@MainActor
final class SpyDelegate: NSObject, BeepingDelegate {
    var beepIdCalls: Int = 0
    var eventCalls: Int = 0
    var lastKey: String?
    var lastEvent: BeepingEvent?

    func beepId(with beepId: String) {
        beepIdCalls += 1
        lastKey = beepId
    }

    func beepingDidReceive(event: BeepingEvent) {
        eventCalls += 1
        lastEvent = event
    }
}

@Suite("Beeping delegate dispatch")
@MainActor
struct BeepingDelegateTests {

    /// Helper that gives every test a fresh delegate wired into the
    /// shared singleton. Tests share the singleton (legacy semantics
    /// preserved) but rotate delegates so they don't interfere.
    private func freshDelegate() -> SpyDelegate {
        let delegate = SpyDelegate()
        Beeping.shared().delegate = delegate
        return delegate
    }

    @Test("EndOk emits both event AND beepId")
    func endOkEmitsEventAndBeepId() {
        let delegate = freshDelegate()
        let event = BeepingEvent(
            status: .endOk,
            key: "12345",
            decodedString: "12345abcd",
            mode: 2,
            timestamp: 99,
            confidence: 0.8,
            confidenceError: 0.1,
            confidenceNoise: 0.2,
            receivedBeepsVolume: -5
        )

        Beeping.shared()._injectEventForTesting(event)

        #expect(delegate.eventCalls == 1)
        #expect(delegate.beepIdCalls == 1)
        #expect(delegate.lastEvent?.status == .endOk)
        #expect(delegate.lastEvent?.key == "12345")
        #expect(delegate.lastKey == "12345")
        #expect(delegate.lastEvent?.decodedString != nil)
    }

    @Test("EndBad emits only event, no beepId")
    func endBadEmitsOnlyEvent() {
        let delegate = freshDelegate()
        let event = BeepingEvent(
            status: .endBad,
            key: nil,
            decodedString: nil,
            mode: 1,
            timestamp: 0,
            confidence: 0.2,
            confidenceError: 0,
            confidenceNoise: 0,
            receivedBeepsVolume: 0
        )

        Beeping.shared()._injectEventForTesting(event)

        #expect(delegate.eventCalls == 1)
        #expect(delegate.beepIdCalls == 0)
        #expect(delegate.lastEvent?.status == .endBad)
    }

    @Test("Start emits only event, no beepId")
    func startEmitsEventOnly() {
        let delegate = freshDelegate()
        let event = BeepingEvent(
            status: .start,
            key: nil,
            decodedString: nil,
            mode: 0,
            timestamp: 0,
            confidence: 0,
            confidenceError: 0,
            confidenceNoise: 0,
            receivedBeepsVolume: 0
        )

        Beeping.shared()._injectEventForTesting(event)

        #expect(delegate.eventCalls == 1)
        #expect(delegate.beepIdCalls == 0)
        #expect(delegate.lastEvent?.status == .start)
    }

    @Test("EndOk with nil key skips beepId (defensive)")
    func endOkWithNilKeySkipsBeepId() {
        let delegate = freshDelegate()
        let event = BeepingEvent(
            status: .endOk,
            key: nil,  // legacy edge case: status endOk but key is nil
            decodedString: nil,
            mode: 0,
            timestamp: 0,
            confidence: 0,
            confidenceError: 0,
            confidenceNoise: 0,
            receivedBeepsVolume: 0
        )

        Beeping.shared()._injectEventForTesting(event)

        #expect(delegate.eventCalls == 1)
        #expect(delegate.beepIdCalls == 0)
    }
}
